test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+# already exists it returns the name
+# Parameters:
+# - PATH: the path where the directory will be created 
+create_dir() { 
+	
+	local path=$1
+	local hashed_name
+	local directory
+	validate_path "$path"
+	
+	# most directories have the same hash, so we only need one directory
+	# in the files directory, if there is already one just return the path/name
+	directory=$(find "$path"/* -type d 2> /dev/null)
+	if [ ! "$directory" ]; then
+		sudo mkdir "$path"/testdir
+		hashed_name=$(sudo "$SWUPD" hashdump "$path"/testdir 2> /dev/null)
+		sudo mv "$path"/testdir "$path"/"$hashed_name"
+		# since tar is all we use, create a tar for the new dir
+		create_tar "$path"/"$hashed_name"
+		directory="$path"/"$hashed_name"
+	fi
+	echo "$directory"
+
+}
+
+# Generates a file with a hashed name in the specified path
+# Parameters:
+# - PATH: the path where the file will be created    
+create_file() {
+ 
+	local path=$1
+	local hashed_name
+	validate_path "$path"
+
+	generate_random_content | sudo tee "$path/testfile" > /dev/null
+	hashed_name=$(sudo "$SWUPD" hashdump "$path"/testfile 2> /dev/null)
+	sudo mv "$path"/testfile "$path"/"$hashed_name"
+	# since tar is all we use, create a tar for the new file
+	create_tar "$path"/"$hashed_name"
+	echo "$path/$hashed_name"
+
+}
+
+# Creates a symbolic link with a hashed name to the specified file in the specified path.
+# If no existing file is specified to point to, a new file will be created and pointed to
+# by the link.
+# Parameters:
+# - PATH: the path where the symbolic link will be created
+# - FILE: the path to the file to point to 
+create_link() { 
+
+	local path=$1
+	local pfile=$2
+	local hashed_name
+	validate_path "$path"
+	
+	# if no file is specified, create one
+	if [ -z "$pfile" ]; then
+		pfile=$(create_file "$path")
+	else
+		validate_item "$pfile"
+	fi
+	sudo ln -rs "$pfile" "$path"/testlink
+	hashed_name=$(sudo "$SWUPD" hashdump "$path"/testlink 2> /dev/null)
+	sudo mv "$path"/testlink "$path"/"$hashed_name"
+	echo "$path/$hashed_name"
+
+}
+
+# Creates a tar for the specified item in the same location
+# Parameters:
+# - ITEM: the relative path to the item (file, directory, link, manifest)
+create_tar() {
+
+	local item=$1
+	local path
+	local item_name
+	validate_item "$item"
+
+	path=$(dirname "$(realpath "$item")")
+	item_name=$(basename "$item")
+	# if the item is a directory exclude its content when taring
+	if [ -d "$item" ]; then
+		sudo tar -C "$path" -cf "$path"/"$item_name".tar --exclude="$item_name"/* "$item_name"
+	else
+		sudo tar -C "$path" -cf "$path"/"$item_name".tar "$item_name"
+	fi
+
+}
+
+# Creates an empty manifest in the specified path
+# Parameters:
+# - PATH: the path where the manifest will be created
+# - BUNDLE_NAME: the name of the bundle which this manifest will be for
+create_manifest() {
+
+	local path=$1
+	local name=$2
+	local version
+	validate_path "$path"
+	validate_param "$name"
+
+	version=$(basename "$path")	
```

Comment:
I think this would be better structured as
```
     {
           printf 'MANIFEST\t3\n'
           printf 'version:\t%s\n' "$version"
           ...
     }  | sudo tee "$path/Manifest.$name" >/dev/null
     echo "$path/Manifest.$name"
```

Reply:
I didn't know of this method, I'll update the code. ;)